app/src/main/java/com/eis/geoCalendar/app/GenericTimedEventManager.java

```diff
@@ -0,0 +1,81 @@
+package com.eis.geoCalendar.app;
+
+import com.eis.geoCalendar.events.EventDatabase;
+import com.eis.geoCalendar.timedEvents.DateTime;
+import com.eis.geoCalendar.timedEvents.TimedEvent;
+import com.eis.geoCalendar.timedEvents.TimedEventManager;
+
+import java.util.ArrayList;
+
+/**
+ * This class manages all scheduled events.
+ *
+ * @param <E> The generic timed event.
+ * @author Francesco Bau' helped by Luca Crema.
+ * @version 1.0
+ * @see com.eis.geoCalendar.app.GenericEventManager
+ * @see com.eis.geoCalendar.timedEvents.TimedEvent
+ * @see com.eis.geoCalendar.timedEvents.TimedEventManager
+ * @since 23/12/2019
+ */
+public class GenericTimedEventManager<E extends TimedEvent> extends GenericEventManager<E> implements TimedEventManager<E> {
+
+    /**
+     * Main constructor, needed to acquire the database of the events.
+     *
+     * @param database The database of the events.
+     */
+    public GenericTimedEventManager(EventDatabase<E> database) {
+        super(database);
+    }
+
+    /**
+     * Searches for all events that have a time earlier than the parameter time
+     *
+     * @param time The time upper bound.
+     * @return An {@link ArrayList} of events with time smaller than {@code time} if there is any, an empty array otherwise.
+     */
+    @Override
+    public ArrayList<E> getEventsBeforeTime(DateTime time) {
```

Comment:
Since you're checking for an event hapened before a given time, and since you have a function that checks for events in a range, you could unify these function by making this one call `getEventsBetweenTime` with `endTime = time` and `beginTime` being the first possible time (which I think it's 01/01/1970)

Reply:
Very nice idea indeed... So basically getEventsBeforeTime and getEventsAfterTime call getEventsBetweenTime, with a default BEGIN_TIME ?

But you also just made me think about one issue: we wanted this to be as generic as possible, so basically, if it's a Calendar, we might want to save something from the past, such as memorable events happened before 1970s... (almost like a Diary / Album with location)
Or maybe an historian / archeologist wants to save the exact time and location of a certain event, or for the origin of a certain object in a museum...
What do you think about it?
But I'm gonna ask if I can do this edit, anyway...

Reply:
I've just seen this fun fact:
Date date = new Date(Long.MIN_VALUE);
returns
Sun Dec 02 22:47:04 BDT 292269055
So yeah, there's indeed a minimum date already! 👍

Reply:
> we might want to save something from the past, such as memorable events happened before 1970s

Well if you want to do this then you can't use java.util.Date, since the first possible date with that is "January 1, 1970, 00:00:00 GMT." as per official documentation.
Also I don't think you can do it with anything earlier than this date since it's basically a standard for dates, and creating your own standard is next to impossible (because of how complicated it is, just look at this: https://www.youtube.com/watch?v=-5wpm-gesOY )
So... better not to I guess

Reply:
Oh I see...

Reply:
Thank you for telling me this 👍

Reply:
Is it better now?

Reply:
Much Better ❤️